### PR TITLE
NIFI-748 Fixed logic around handling partial query results from prove…

### PR DESCRIPTION
**Pull request by olegz**
…nance repository

- Ensured that failures derived form correlating Document to its actual provenance event do fail the entire query and produce partial results with warning messages
- Refactored DocsReader.read() operation.
- Added test to validate two conditions where the such failures could occur

### CODE REVIEW

File: nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java [hunk 1]
```diff
@@ -47,9 +46,6 @@
 public class DocsReader {
     private final Logger logger = LoggerFactory.getLogger(DocsReader.class);
 
-    public DocsReader(final List<File> storageDirectories) {
-    }
-
```
Comment (trkurc) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java]: Removing this public method is a breaking change.

Comment (olegz) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java]: Considering that this class does not appear to be intended as public API I still believe it is the right thing to do since the constructor argument isn't used. 
In fact, need to look if we can make this class package private.

File: nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java [hunk 2]
```diff
@@ -100,101 +96,61 @@ private ProvenanceEventRecord getRecord(final Document d, final RecordReader rea
...
-                    } else {
-                        logger.debug("Opening log file {}", storageFilename);
-
-                        logFileCount++;
-                        if (reader != null) {
-                            reader.close();
-                        }
+        Set<ProvenanceEventRecord> matchingRecords = new LinkedHashSet<>();
+        if (retrievalCount.get() >= maxResults) {
+            return matchingRecords;
```
Comment (trkurc) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java]: Out of curiousity, why was this moved and changed? It appears to be same functionality as before, when returning emptySet, but now requires an object creation

Comment (olegz) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java]: I think I'd agree. Just felt a bit cleaner

File: nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java [hunk 3]
```diff
@@ -100,101 +96,61 @@ private ProvenanceEventRecord getRecord(final Document d, final RecordReader rea
...
+                            eventsReadThisFile++;
+
+                            if (retrievalCount.incrementAndGet() >= maxResults) {
+                                break;
                             }
                         }
                     }
-                } finally {
-                    lastStorageFilename = storageFilename;
+                } catch (Exception e) {
```
Comment (trkurc) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java]: Did you consider breaking out the different types exceptions here, which could help with diagnosis? Also, although unlikely (and possibly not plausible), the reader could throw an exception on close. Did you consider a closeQuietly instead of a try-with-resources?

File: nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/LuceneUtil.java [hunk 4]
```diff
@@ -160,4 +162,25 @@ public int compare(final Document o1, final Document o2) {
...
+    /**
+     * Will group documents based on the {@link FieldNames#STORAGE_FILENAME}.
+     *
+     * @param documents
+     *            list of {@link Document}s
+     * @return a {@link Map} of document groups with
+     *         {@link FieldNames#STORAGE_FILENAME} as key and {@link List} of
+     *         {@link Document}s as value.
+     */
+    public static Map<String, List<Document>> groupDocsByStorageFileName(final List<Document> documents) {
```
Comment (trkurc) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/LuceneUtil.java]: I like this change. It greatly simplifies the code

File: nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java (hunk 3 above)
Comment (olegz) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java]: Good point, but at the time I don't feel the need since we don't have specific requirements for different behavior based on exception type. As for the close() I'd be interested to see what's wrong. May be in the future we'll earn enough not to care, then we can change to close quietly

File: nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java [hunk 5]
```diff
@@ -100,101 +96,61 @@ private ProvenanceEventRecord getRecord(final Document d, final RecordReader rea
...
+        return record;
+    }
 
     public Set<ProvenanceEventRecord> read(final List<Document> docs, final Collection<Path> allProvenanceLogFiles,
-        final AtomicInteger retrievalCount, final int maxResults, final int maxAttributeChars) throws IOException {
-        if (retrievalCount.get() >= maxResults) {
-            return Collections.emptySet();
-        }
-
-        LuceneUtil.sortDocsForRetrieval(docs);
```
Comment (trkurc) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java]: It appears this was being done 'for performance'. Did you run performance tests before and after?

Comment (olegz) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java]: Hmm, I don't think so. Based on my observation it was done to group files based on storage file name and sorting had that sideeffect, so in a way the additional sorting is a bit of an overkill here, hence refactoring and new utility operation.

File: nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java (hunk 1 above)
Comment (joewitt) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java]: Should consult the versioning guide to see if it addresses how to handle this.  We need to be very explicit about which parts of the codebase are public and designed for extension versus those which the community should be free to evolve .  if the guide doesn't cover this case let's solve it.

File: nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java (hunk 2 above)
Comment (joewitt) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java]: We def need to be careful in here.  Performance and implications on GC in general are critical in anything prov related.

File: nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java (hunk 1 above)
Comment (trkurc) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java]: I'm like 95% with @olegz  on this. I believe it is implausible that someone would be calling this method and likely needed a comment that it was public for some other reason than it was a public API. That 5% not with @olegz is adding deprecation tag adds a tiny smidge of tech debt, which would be swept up when we investigate all the deprecated methods for 1.0.0

File: nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java (hunk 2 above)
Comment (trkurc) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java]: @joewitt - if this causes GC problems we've got other _major_ issues. I was just bit surprised a change was made that in my opinion was marginally and subjectively a step backwards.

File: nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java (hunk 5 above)
Comment (trkurc) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java]: I base this on  the comment from the sort method:
 // sort by filename and then file offset so that we can retrieve the records efficiently

Comment (olegz) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java]: Tony, I'll give it another look

File: nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java (hunk 1 above)
Comment (joewitt) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java]: We have designed flexibility in for a reason.  Deprecation is appropriate for the public portions of code designed for extension and/or modification.  For everything else let's be sure it is clear and documented what is fair game for the community to evolve.  If it isn't clear then we should err on the side of deprecation then make it clear.  We need to have agility behind the curtain.

Comment (trkurc) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java]: 100% agree. Other projects use annotations to warn folks of risk when using public methods that aren't extension points (e.g. @PublicForTests or @Experimental), but I'd settle for a comment.

Comment (olegz) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java]: I agree that depreciation is the right thing to do in general cases like this. But given that the project is pre-1.0 and obvious intent of this class not being a public API I felt removing it would be ok. Still do but will honor the majority opinion and will deprecate.

Comment (joewitt) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java]: oleg i was entirely in agreement with how you had it for this case (removal).  Tony was only 5% in disagreement.  Of those that weighed in you have 195% support ;-)

Comment (apiri) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java]: I think there has been a "handshake agreement" on the nature of the framework code versus the extension developer facing API.  Having said that, I was not able to find any documentation that indicates that is the case.  As a result, I think I am a similar 95/5 split as Tony.  As a result, I think the "right" thing given the circumstances is to tag it up as Deprecated and remove at 1.x.  However, I think it's likely a fair assumption that this is clear.  Either way, we need to get the handshake agreement in a tangible form some place.  I'll make a ticket and please add on/enhance as needed.

**Comment by apiri** [comment]
Grabbed the PR locally and scoped it out.  Nothing substantive to add beyond those comments here already.  Performed some querying of a locally running instance with the new patch applied and didn't notice any tremendous performance impacts.  Will continue to run as this patch goes through it's final paces.

**Comment by olegz** [comment]
@trkurc @joewitt @apiri 
Guys, please see the latest commit. Didn't squash it, so its easier to read and see what's been addressed. In summary:
1. Since based on the latest comment from Joe it appears that we all agree that DocReader is not really public, i kept the dead constructor out and also made DocReader package private.
2. Based on Tony's point added Document sorting logic back. At least it will ensure that previous behavior is maintained. 
See commit message for more details

**Comment by trkurc** [comment]
I'm mostly okay with package private. On the off chance someone was using this API it will be marginally more challenging to adapt to this change, but is possible.

**Comment by markap14** [comment]
@trkurc  Personally, I have exactly 0 qualms about changing it to package private. If I choose to take some random util class from a release of Apache Tomcat, for example, and depended on it, I should certainly not be surprised if that class changes from release to release (including incremental releases) - I don't think this is any different.

File: nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java (hunk 1 above)
Comment (markap14) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java]: @apiri The versioning scheme is defined at https://cwiki.apache.org/confluence/display/NIFI/Version+Scheme - it essentially says that all extension points that adhere to API version X.Y.* must also work in version X.(Y+1).*. I.e., minor changes will not break backward compatibility of extension point API or those objects that can be reached by extension points via the API.
